File: Venkatesh/DFS-BFS/ShortestPath.py
def shortest_path(graph, start, end):
    queue = [(start, [start])]
    shortest = None
    while queue:
        (vertex, path) = queue.pop(0)
        for next in set(graph[vertex]) - set(path):
            if next == end:
                if shortest is None or len(path) + 1 < len(shortest):
                    shortest = path + [next]
            else:
                queue.append((next, path + [next]))
    return shortest

nodes = ['A', 'B', 'C', 'D', 'E']
edges = [('A', 'B'), ('A', 'C'), ('B', 'D'), ('C', 'E'), ('D', 'E')]
start_node = 'A'
end_node = 'E'

graph = {node: set() for node in nodes}
for start_edge, end_edge in edges:
    graph[start_edge].add(end_edge)

print(shortest_path(graph, start_node, end_node))

File: Venkatesh/DFS-BFS/ShortestPath.java
import java.util.*;

public class ShortestPath {

    public static List<String> shortestPath(Map<String, Set<String>> graph, String start, String end) {
        Queue<List<String>> queue = new ArrayDeque<>();
        queue.add(Arrays.asList(start));
        List<String> shortest = null;

        while (!queue.isEmpty()) {
            List<String> path = queue.poll();
            String vertex = path.get(path.size() - 1);

            for (String next : graph.get(vertex)) {
                if (!path.contains(next)) {
                    List<String> newPath = new ArrayList<>(path);
                    newPath.add(next);

                    if (next.equals(end)) {
                        if (shortest == null || newPath.size() < shortest.size()) {
                            shortest = newPath;
                        }
                    } else {
                        queue.add(newPath);
                    }
                }
            }
        }
        return shortest;
    }

    public static void main(String[] args) {
        String[] nodes = {"A", "B", "C", "D", "E"};
        String[][] edges = {{"A", "B"}, {"A", "C"}, {"B", "D"}, {"C", "E"}, {"D", "E"}};
        String startNode = "A";
        String endNode = "E";

        Map<String, Set<String>> graph = new HashMap<>();
        for (String node : nodes) {
            graph.put(node, new HashSet<>());
        }
        for (String[] edge : edges) {
            graph.get(edge[0]).add(edge[1]);
        }

        List<String> path = shortestPath(graph, startNode, endNode);
        System.out.println(path);
    }
}
